File: api-python/services/whois_service.py
import whois

async def fetch_whois_data(domain: str):
    """
    Fetch WHOIS records for the given domain.
    :param domain: Domain name
    :return: WHOIS records
    """
    try:
        whois_data = whois.whois(domain)
        if whois_data is None:
            raise Exception("WHOIS data not found")
        return whois_data
    except Exception as e:
        raise Exception(f"Failed to fetch WHOIS data: {str(e)}")

File: api/src/service/whois.service.ts
import { Injectable } from '@nestjs/common';
import * as whois from 'whois-json';

@Injectable()
export class WhoisService {
  async lookup(domain: string): Promise<Record<string, any>> {
    if (!domain || domain.trim() === '') {
      throw new Error('Domain cannot be empty.');
    }

    try {
      const whoisData = await whois(domain, { follow: 3 }); // Follow up to 3 whois redirects
      return whoisData;
    } catch (error) {
      throw new Error('Failed to fetch WHOIS data: ' + error.message);
    }
  }
}
